### PR TITLE
Install Script v2

### PR DESCRIPTION
**Pull request by XDream8**
# What is done
- Code improvents/cleanup
- New option management system `./install.sh --install-path=/usr/bin --dry-run`
- Automatic detection of $install_path directory group using stat, if group = root script will use $sudo
- Options
```
    --root,     -r
    --noroot,   -nr
    --dry-run,  -dr
    --nocolors, -nc
    --headless, -y
    --help,     -h
```

# TODO
- [x] add headless install back
- [x] format the script using shfmt

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -136,13 +136,14 @@ fi
...
 	if check_dep "find" && [ "$(find /usr/bin /usr/local/bin $HOME/.local/bin -type f -iname 'fm6000')" ]; then
 		out "${RED}it seems fm6000 is already installed but continuing anyway${NC}"
 	fi
 
-	notset "$dryrun" && {
-		printf '%b' "${YELLOW}"
-		notset "$headless" && read -p "Move the script to $install_path [$require_text]? (y/N) " ans
+	notset "$dryrun" && notset "$headless" && {
+		printf '%b' "${YELLOW}Move the script to $install_path [${RED}$require_text${YELLOW}]? (y/N) "
+		read -r ans
```
Comment (XDream8) [inline install.sh:146]: support for posix read

File: install.sh [hunk 2]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+	printf '%b\n' "$@"
+}
+
+notset() {
+	case $1 in '') return 0 ;; *) return 1 ;; esac
+}
+
+check_dep() {
+	if [ "$(command -v "$1")" ]; then return 0; else return 1; fi
+}
```
Comment (XDream8) [inline install.sh:28]: create functions for commonly used things

File: install.sh [hunk 3]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+ ${RED}--nocolors${NC},${RED} -nc    ${GREEN}Do not print colored text${NC}
+ ${RED}--root${NC},${RED}     -r     ${GREEN}Use root privileges(it is automatically detected but still can be used)${NC}
+ ${RED}--noroot${NC},${RED}   -nr    ${GREEN}Do not use root privileges(it is automatically detected but still can be used)${NC}
+ ${RED}--dry-run${NC},${RED}  -dr    ${GREEN}Dry run fm6000${NC}
+ ${RED}--headless${NC},${RED} -y     ${GREEN}install headless(without prompting)${NC}
+"
+		return 1
+		;;
+	esac
+done
```
Comment (XDream8) [inline install.sh:69]: New option management system

File: install.sh [hunk 4]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+# make script compatible with env variables
+
+## set install_path
+notset "$install_path" && {
+	if [ -x "$HOME/.local/bin" ]; then
+		install_path=$HOME/.local/bin
+	elif [ -x "/usr/local/bin" ]; then
+		install_path=/usr/local/bin
+	fi
+}
```
Comment (XDream8) [inline install.sh:80]: Use one of the existant install path, if it is not set with options

**default install paths**
- $HOME/.local/bin
- /usr/local/bin

File: install.sh [hunk 5]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+	[ ! -x "$install_path" ] && {
+		out "${RED}$install_path does not exist${NC}"
+		return 1
+	}
+	check_dep "stat" && {
+		if [ "$(stat -c "%G" $install_path)" = "root" ]; then
+			notset "$root" && root=1
+		fi
+	}
+}
```
Comment (XDream8) [inline install.sh:93]: Check the group of install path

if group equals to root, root privileges will be used

File: install.sh [hunk 6]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
 else
-    url="https://raw.githubusercontent.com/anhsirk0/fetch-master-6000/master/fm6000.pl"
-    if [ "$(command -v curl)" ]; then
-        printf '%b\n' "${BLUE}Downloading the script${NC}"
-        curl $url -o fm6000
-    else
-        printf '%b\n' "${RED}curl is required${NC}" && exit 1
-    fi
+	require_text="root required"
 fi
```
Comment (XDream8) [inline install.sh:104]: set root requirement text

File: install.sh [hunk 7]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+		curl $url -o fm6000
+	else
+		out "${RED}curl is required${NC}" && return 1
+	fi
+fi
+
+if ! notset "$root"; then
+	## if user is root do not set sudo
+	if [ "$(id -u)" -eq "0" ]; then
+		sudo=
```
Comment (XDream8) [inline install.sh:126]: if root user is running the script do not use sudo

File: install.sh [hunk 8]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
-        printf '%b\n' "${YELLOW}Script not moved!${NC}"
-        ./fm6000
-        exit 0
-    fi
+	chmod +x fm6000 && out "${BLUE}Making the script executable : ${GREEN}done"
+
+	# shellcheck disable=SC2086
+	if check_dep "find" && [ "$(find /usr/bin /usr/local/bin $HOME/.local/bin -type f -iname 'fm6000')" ]; then
+		out "${RED}it seems fm6000 is already installed but continuing anyway${NC}"
+	fi
```
Comment (XDream8) [inline install.sh:142]: Check if fm6000 is already available in /usr/bin or /usr/local/bin or $HOME/.local/bin

File: install.sh [hunk 9]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+	}
+
+	case $ans in
+	y | Y | -y | -Y)
+		ans="y"
+		;;
+	*)
+		ans="N"
+		;;
+	esac
```
Comment (XDream8) [inline install.sh:156]: support multiple answers

File: install.sh [hunk 10]
```diff
@@ -17,58 +14,161 @@ CYAN='\033[0;36m'
...
+
+	if [ "$headless" = "1" ] || [ "$ans" = "y" ]; then
+		out "${BLUE}Moving fm6000 to $install_path${NC}"
+		printf '%b' "${CYAN}"
+		$sudo mv -v fm6000 $install_path || {
+			out "${RED}error: $sudo failed${NC}"
+			return 1
+		}
```
Comment (XDream8) [inline install.sh:164]: Better error handling

**Comment by anhsirk0** [comment]
Thanks for all the work. This is great

**Pull request merged**